File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        Map<Integer, Integer> hash = new HashMap<>();
        Stack<Integer> st = new Stack<>();
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);

        for (int i = 0; i < 2 * n; i++) {
            int current = nums[i % n];
            
            while (!st.isEmpty() && nums[st.peek()] < current) {
                res[st.pop()] = current;
            }

            if (i < n) {
                st.push(i);
            }
        }


        return res;
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [-1] * n
        stack = []
        for i in range(2 * n):
            while stack and nums[i % n] > nums[stack[-1]]:
                ans[stack.pop()] = nums[i % n]

            if i < n:
                stack.append(i)
        return ans

        